### PR TITLE
[WIP] Python 3 port

### PR DESCRIPTION
**Pull request by redapple**
Scrapy 1.1 is required for `to_bytes, to_unicode, to_native_str`

Major changes:
- quite a few `zope.interface.implements` --> `zope.interface.implementer`
- introduce a `native_stringify_dict` to help with interpreting URL parameters among others
- `DictMixin` replaced with `MutableMapping` for sqlite backend
- use paths as bytes when registering children resource for Home resource
- explicit UTF-8 encoding for HTTP responses

### CODE REVIEW

**Comment by Digenis** [comment]
The latest scrapy I tested thoroughly with the current stable scrapyd, is 1.0.
Unless scrapyd-1.1 works with scrapy-1.1,
setting the minimum scrapy requirement to 1.1
will mean that one has to upgrade the 2 packages together.

I took a look at the changes using to_bytes, to_unicode, to_native_str
and I think in most of them we are expected to know what type to anticipate.
I'm willing to try opening a PR on top of python3-wip.
Did you try without these utils?

**Comment by redapple** [comment]
@Digenis , what do you mean "without these utils"? with pre-scrapy1.1 utils?

**Comment by Digenis** [comment]
without to_bytes, to_unicode, to_native_str

**Comment by redapple** [comment]
not using them or using std library conversion?

**Comment by redapple** [comment]
Please have a go at a PR against python3-wip

**Comment by redapple** [comment]
@Digenis , as a cheap-and-ugly workaround, we can also copy-paste the 3 utils functions to scrapyd

**Comment by redapple** [comment]
So I tried to simplify the encoding stuff, and ran into an issue with non-ASCII spider names.
Seems listing them is fine, but scheduling them doesn't work even for a simple case and in Python 2 (I will test with current master branch too)
So this needs further work.
DO NOT MERGE

**Comment by redapple** [comment]
FTR, current master branch also fails with non-ASCII spider names, but differently.
When deploying a project with a spider having a non-ASCII name:

spider code:
```
$ cat unicodename/spiders/example.py 
# -*- coding: utf-8 -*-
import scrapy


class ExampleSpider(scrapy.Spider):
    name = u"araña"
    allowed_domains = ["example.com"]
    start_urls = (
        'http://www.example.com/',
    )

    def parse(self, response):
        yield {"url": response.url}
```

scrapyd-deploy:
```
$ scrapyd-deploy 
Packing version 1473330688
Deploying to project "unicodename" in http://localhost:6800/addversion.json
Deploy failed: <urlopen error [Errno 111] Connection refused>
(scrapyd-client.py27) paul@paul-SATELLITE-R830:~/scrapinghub/scrapy/unicodename$ scrapyd-deploy 
Packing version 1473330697
Deploying to project "unicodename" in http://localhost:6800/addversion.json
Server response (200):
{"status": "error", "message": "UnicodeEncodeError: 'ascii' codec can't encode character u'\\xf1' in position 3: ordinal not in range(128)", "node_name": "paul-SATELLITE-R830"}
```

scrapyd console logs:
```
2016-09-08T12:31:37+0200 [_GenericHTTPChannelProtocol,0,127.0.0.1] Unhandled Error
	Traceback (most recent call last):
	  File "/home/paul/.virtualenvs/scrapyd/local/lib/python2.7/site-packages/twisted/web/http.py", line 1845, in allContentReceived
	    req.requestReceived(command, path, version)
	  File "/home/paul/.virtualenvs/scrapyd/local/lib/python2.7/site-packages/twisted/web/http.py", line 766, in requestReceived
	    self.process()
	  File "/home/paul/.virtualenvs/scrapyd/local/lib/python2.7/site-packages/twisted/web/server.py", line 190, in process
	    self.render(resrc)
	  File "/home/paul/.virtualenvs/scrapyd/local/lib/python2.7/site-packages/twisted/web/server.py", line 241, in render
	    body = resrc.render(self)
	--- <exception caught here> ---
	  File "/home/paul/src/scrapyd/scrapyd/webservice.py", line 17, in render
	    return JsonResource.render(self, txrequest)
	  File "/home/paul/src/scrapyd/scrapyd/utils.py", line 19, in render
	    r = resource.Resource.render(self, txrequest)
	  File "/home/paul/.virtualenvs/scrapyd/local/lib/python2.7/site-packages/twisted/web/resource.py", line 250, in render
	    return m(request)
	  File "/home/paul/src/scrapyd/scrapyd/webservice.py", line 79, in render_POST
	    spiders = get_spider_list(project, version=version)
	  File "/home/paul/src/scrapyd/scrapyd/utils.py", line 116, in get_spider_list
	    raise RuntimeError(msg.splitlines()[-1])
	exceptions.RuntimeError: UnicodeEncodeError: 'ascii' codec can't encode character u'\xf1' in position 3: ordinal not in range(128)
```

**Comment by redapple** [comment]
Regarding non-ASCII spider names (in Python 2), this [is not supported in Scrapy](https://github.com/scrapy/scrapy/issues/2251) so no need to worry too much.

**Comment by Digenis** [comment]
@redapple I tried to reproduce https://github.com/scrapy/scrapyd/pull/179#issuecomment-245558755 on python3
but I get no error.
Does your last comment imply that this branch is ready?

**Comment by redapple** [comment]
hi @Digenis , I haven't looked at this in a while but non-ASCII names are not supposed to work on Python 2, so my tests from that https://github.com/scrapy/scrapyd/pull/179#issuecomment-245556877 are irrelevant.
Maybe the encoding stuff can be simplified still, I'm not sure.

**Comment by redapple** [comment]
@Digenis , let's merge this and work towards a alpha or beta release of a Python 3 compatible Scrapyd.
We did that for Scrapy with Release Candidates and I think it worked not too bad.
what do you think?

File: scrapyd/utils.py [hunk 1]
```diff
@@ -114,7 +114,9 @@ def get_spider_list(project, runner=None, pythonpath=None, version=''):
     if proc.returncode:
         msg = err or out or 'unknown error'
         raise RuntimeError(msg.splitlines()[-1])
```
Comment (Digenis) [inline scrapyd/utils.py]: @redapple,
I think we should also decode (from utf8) here
and then, if we are on python2, again encode the msg to unicode_escape before raising.
@kmike often advised this in scrapy.
Shall I commit it directly to master?
I know that the same "FIXME" comment will apply to this line too.

Comment (Digenis) [inline scrapyd/utils.py]: 3c24f59cb8bd18dc1597a658177597940da97bbe is what I mean

**Pull request merged**